File: src/vfs.js
// MIT License

// Copyright (c) 2022-2024 Tobias Enderle

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

"use strict";

addToLibrary({
  nodejsAccess: function (vfs, filePath, flags, outResult) {
    /* c8 ignore stop */
    let aflags = fs.constants.F_OK;
    if (flags == SQLITE_ACCESS_READWRITE)
      aflags = fs.constants.R_OK | fs.constants.W_OK;
    if (flags == SQLITE_ACCESS_READ) aflags = fs.constants.R_OK;
    try {
      fs.accessSync(UTF8ToString(filePath), aflags);
      setValue(outResult, 1, "i32");
    } catch {
      setValue(outResult, 0, "i32");
    }
    return SQLITE_OK;
  },
  nodejsFullPathname: function (vfs, relPath, sizeFullPath, outFullPath) {
    const full = path.resolve(UTF8ToString(relPath));
    stringToUTF8(full, outFullPath, sizeFullPath);
    return full.length < sizeFullPath ? SQLITE_OK : SQLITE_CANTOPEN;
  },
  nodejsWrite: function (fi, buffer, bytes, offset) {
    try {
      const bytesWritten = fs.writeSync(
        _fd(fi),
        HEAPU8.subarray(buffer, buffer + bytes),
        0,
        bytes,
        _safeInt(offset)
      );
      return bytesWritten != bytes ? SQLITE_IOERR_WRITE : SQLITE_OK;
    } catch {
      return SQLITE_IOERR_WRITE;
    }
  },
  nodejsSync: function (fi, flags) {
    try {
      fs.fsyncSync(_fd(fi));
    } catch {
      return SQLITE_IOERR_FSYNC;
    }
    return SQLITE_OK;
  },
  nodejsClose: function (fi) {
    _nodejsUnlock(fi, SQLITE_LOCK_NONE);
    try {
      fs.closeSync(_fd(fi));
    } catch {
      return SQLITE_IOERR_CLOSE;
    }
    return SQLITE_OK;
  },
  nodejsRead: function (fi, outBuffer, bytes, offset) {
    const buf = HEAPU8.subarray(outBuffer, outBuffer + bytes);
    let bytesRead;
    try {
      bytesRead = fs.readSync(_fd(fi), buf, 0, bytes, offset);
    } catch {
      return SQLITE_IOERR_READ;
    }
    if (bytesRead == bytes) {
      return SQLITE_OK;
    } else if (bytesRead >= 0) {
      if (bytesRead < bytes) {
        try {
          buf.fill(0, bytesRead);
        } catch {
          return SQLITE_IOERR_READ;
        }
      }
      return SQLITE_IOERR_SHORT_READ;
    }
    return SQLITE_IOERR_READ;
  },
  nodejsDelete: function (vfs, filePath, dirSync) {
    const pathStr = UTF8ToString(filePath);
    try {
      fs.unlinkSync(pathStr);
    } catch (err) {
      if (err.code != "ENOENT") return SQLITE_IOERR_DELETE;
    }
    if (dirSync) {
      let fd = -1;
      try {
        fd = fs.openSync(path.dirname(pathStr), "r");
        fs.fsyncSync(fd);
      } catch {
        return SQLITE_IOERR_FSYNC;
      } finally {
        try {
          fs.closeSync(fd);
        } catch {
          return SQLITE_IOERR_FSYNC;
        }
      }
    }
    return SQLITE_OK;
  },
  nodejsRandomness: function (vfs, bytes, outBuffer) {
    const buf = HEAPU8.subarray(outBuffer, outBuffer + bytes);
    crypto.randomFillSync(buf);
    return bytes;
  },
  nodejsTruncate: function (fi, size) {
    try {
      fs.ftruncateSync(_fd(fi), _safeInt(size));
    } catch {
      return SQLITE_IOERR_TRUNCATE;
    }
    return SQLITE_OK;
  },
  nodejsFileSize: function (fi, outSize) {
    try {
      setValue(outSize, fs.fstatSync(_fd(fi)).size, "i64");
    } catch {
      return SQLITE_IOERR_FSTAT;
    }
    return SQLITE_OK;
  },
  nodejsLock: function (fi, level) {
    // level is never SQLITE_LOCK_NONE
    if (!_isLocked(fi)) {
      try {
        fs.mkdirSync(`${_path(fi)}.lock`);
      } catch (err) {
        return err.code == "EEXIST" ? SQLITE_BUSY : SQLITE_IOERR_LOCK;
      }
      _setLocked(fi, true);
    }
    return SQLITE_OK;
  },
  nodejsUnlock: function (fi, level) {
    // level is either SQLITE_LOCK_NONE or SQLITE_LOCK_SHARED
    if (level == SQLITE_LOCK_NONE && _isLocked(fi)) {
      try {
        fs.rmdirSync(`${_path(fi)}.lock`);
      } catch (err) {
        if (err.code != "ENOENT") return SQLITE_IOERR_UNLOCK;
      }
      _setLocked(fi, false);
    }
    return SQLITE_OK;
  },
  nodejsCheckReservedLock: function (fi, outResult) {
    try {
      fs.accessSync(`${_path(fi)}.lock`, fs.constants.F_OK);
      setValue(outResult, 1, "i32");
    } catch {
      setValue(outResult, 0, "i32");
    }
    return SQLITE_OK;
  },
  nodejs_max_path_length: function () {
    return process.platform == "win32" ? 260 : 4096;
  },
  nodejs_open: function (filePath, flags, mode) {
    let oflags = 0;
    if (flags & SQLITE_OPEN_EXCLUSIVE) oflags |= fs.constants.O_EXCL;
    if (flags & SQLITE_OPEN_CREATE) oflags |= fs.constants.O_CREAT;
    if (flags & SQLITE_OPEN_READONLY) oflags |= fs.constants.O_RDONLY;
    if (flags & SQLITE_OPEN_READWRITE) oflags |= fs.constants.O_RDWR;

    try {
      return fs.openSync(UTF8ToString(filePath), oflags, mode);
    } catch {
      return -1;
    }
    /* c8 ignore start */
  },
});

File: src/vfs.c
// MIT License

// Copyright (c) 2022-2024 Tobias Enderle

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

// This VFS is loosely based on SQLite's test_demovfs.c

#include <string.h>
#include <sys/time.h>
#include <unistd.h>

#include "sqlite3.h"

typedef struct NodeJsFile NodeJsFile;
struct NodeJsFile {
  sqlite3_file base;  // Base class. Must be first.
  int fd;             // File descriptor
  int isLocked;
  const char *path;
};

extern int nodejsWrite(sqlite3_file *, const void *, int, sqlite_int64);
extern int nodejsClose(sqlite3_file *);
extern int nodejsRead(sqlite3_file *, void *, int, sqlite_int64);
extern int nodejsSync(sqlite3_file *, int);
extern int nodejsDelete(sqlite3_vfs *, const char *, int);
extern int nodejsFullPathname(sqlite3_vfs *, const char *, int, char *);
extern int nodejsAccess(sqlite3_vfs *, const char *, int, int *);
extern int nodejsRandomness(sqlite3_vfs *, int, char *);
extern int nodejsTruncate(sqlite3_file *, sqlite_int64);
extern int nodejsFileSize(sqlite3_file *, sqlite_int64 *);
extern int nodejsLock(sqlite3_file *, int);
extern int nodejsUnlock(sqlite3_file *, int);
extern int nodejsCheckReservedLock(sqlite3_file *, int *);

extern int nodejs_open(const char *, int, int);
extern int nodejs_max_path_length();

/*
** No xFileControl() verbs are implemented by this VFS.
*/
static int nodejsFileControl(sqlite3_file *pFile, int op, void *pArg) {
  return SQLITE_NOTFOUND;
}

/*
** The xSectorSize() and xDeviceCharacteristics() methods. These two
** may return special values allowing SQLite to optimize file-system
** access to some extent. But it is also safe to simply return 0.
*/
static int nodejsSectorSize(sqlite3_file *pFile) { return 0; }
static int nodejsDeviceCharacteristics(sqlite3_file *pFile) { return 0; }

static int nodejsOpen(
    sqlite3_vfs *pVfs,    // VFS
    const char *zName,    // File to open, or NULL for a temp file
    sqlite3_file *pFile,  // Pointer to NodeJsFile struct to populate
    int flags,            // Input SQLITE_OPEN_XXX flags
    int *pOutFlags        // Output SQLITE_OPEN_XXX flags (or NULL)
) {
  static const sqlite3_io_methods nodejsio = {
      1,                           // iVersion
      nodejsClose,                 // xClose
      nodejsRead,                  // xRead
      nodejsWrite,                 // xWrite
      nodejsTruncate,              // xTruncate
      nodejsSync,                  // xSync
      nodejsFileSize,              // xFileSize
      nodejsLock,                  // xLock
      nodejsUnlock,                // xUnlock
      nodejsCheckReservedLock,     // xCheckReservedLock
      nodejsFileControl,           // xFileControl
      nodejsSectorSize,            // xSectorSize
      nodejsDeviceCharacteristics  // xDeviceCharacteristics
  };

  NodeJsFile *p = (NodeJsFile *)pFile;
  memset(p, 0, sizeof(NodeJsFile));

  // Temp files not supported.
  if (zName == NULL) return SQLITE_IOERR;

  p->fd = nodejs_open(zName, flags, 0600);
  if (p->fd < 0) return SQLITE_CANTOPEN;
  if (pOutFlags) *pOutFlags = flags;
  p->base.pMethods = &nodejsio;
  p->path = zName;
  return SQLITE_OK;
}

/*
** This VFS does not support loading extensions
*/
static void *nodejsDlOpen(sqlite3_vfs *pVfs, const char *zPath) { return NULL; }
static void nodejsDlError(sqlite3_vfs *pVfs, int nByte, char *zErrMsg) {
  sqlite3_snprintf(nByte, zErrMsg, "Loadable extensions are not supported");
}
static void (*nodejsDlSym(sqlite3_vfs *pVfs, void *pH, const char *z))(void) {
  return 0;
}
static void nodejsDlClose(sqlite3_vfs *pVfs, void *pHandle) {}

static int nodejsSleep(sqlite3_vfs *pVfs, int nMicro) {
  if (nMicro >= 1000000) sleep(nMicro / 1000000);
  if (nMicro % 1000000) usleep(nMicro % 1000000);
  return nMicro;
}

static int nodejsCurrentTimeInt64(sqlite3_vfs *pVfs, sqlite3_int64 *piNow) {
  static const sqlite3_int64 unixEpoch = 24405875 * (sqlite3_int64)8640000;
  struct timeval sNow;
  gettimeofday(&sNow, 0);
  *piNow = unixEpoch + 1000 * (sqlite3_int64)sNow.tv_sec + sNow.tv_usec / 1000;
  return SQLITE_OK;
}

SQLITE_API int sqlite3_os_init(void) {
  static sqlite3_vfs nodejsvfs = {
      2,                      // iVersion
      sizeof(NodeJsFile),     // szOsFile
      -1,                     // mxPathname --- set below
      NULL,                   // pNext
      "nodejs",               // zName
      NULL,                   // pAppData
      nodejsOpen,             // xOpen
      nodejsDelete,           // xDelete
      nodejsAccess,           // xAccess
      nodejsFullPathname,     // xFullPathname
      nodejsDlOpen,           // xDlOpen
      nodejsDlError,          // xDlError
      nodejsDlSym,            // xDlSym
      nodejsDlClose,          // xDlClose
      nodejsRandomness,       // xRandomness
      nodejsSleep,            // xSleep
      NULL,                   // xCurrentTime
      NULL,                   // xGetLastError
      nodejsCurrentTimeInt64  // xCurrentTimeInt64
  };
  nodejsvfs.mxPathname = nodejs_max_path_length();
  sqlite3_vfs_register(&nodejsvfs, 1);
  return SQLITE_OK;
}

SQLITE_API int sqlite3_os_end(void) { return SQLITE_OK; }
